### DISCUSSION TITLE
Identifying streams coming from different clients

### DISCUSSION THREAD

**Discussion by ZE-Ridge-Run** [microsoft/msquic#2829] [Q&A] [answered]
Hello,
Suppose I have 2 clients (A and B) on different threads, both sending data to the same port on a server. On that server I see 2 connections and 2 streams, Is it possible to set some kind of ID to the connection/stream on the client, so that on the server I can identify which stream comes form A and which stream comes from B?

**Comment by anrossi** [comment, marked as answer]
Hi,
So you're trying to figure out how to tell two connections apart on the server? You'll have to send some data in a stream on each connection which identifies that client to the server, and then store this data in the context pointer you set on the connection, and in the context pointer set on each stream.  One way to do this is to have a single context struct for each connection, and then set that context pointer on each stream as well. When the connection is closed, then you free the context struct for the connection.

**Reply by nibanks** [reply]
As Anthony said, the way to handle this is to associate your own "context" object to each stream and connection involved. Then you can use your object to determine the relevant state to make your decision. When you call [ConnectionOpen](https://github.com/microsoft/msquic/blob/main/docs/api/ConnectionOpen.md) or [StreamOpen](https://github.com/microsoft/msquic/blob/main/docs/api/StreamOpen.md) locally, they take the `Context` pointer as argument. If the peer opens the connection or stream, you will be required to call [SetCallbackHandler](https://github.com/microsoft/msquic/blob/main/docs/api/SetCallbackHandler.md) to set it. Then, every future callback you get will have the `Context` passed back to you for you to use as necessary.

**Reply by ZE-Ridge-Run** [reply]
thank you very much, both of you! :)
